slangpy/torchintegration/torchtensormarshall.py

```diff
@@ -1 +1,272 @@
+) -> Optional[SlangType]:
+    scalar_type = _torch_to_scalar_type.get(torch_dtype)
+    if scalar_type is None:
+        return None
+    return layout.scalar_type(scalar_type)
+
+
+def get_storage(context: CallContext, element_count: int, struct_size: int) -> Buffer:
+    return context.device.create_buffer(
+        size=element_count * struct_size,
+        struct_size=struct_size,
+        usage=BufferUsage.shared | BufferUsage.unordered_access | BufferUsage.shader_resource,
+    )
+
+
+class TensorRefMarshall(TensorMarshall):
+    def __init__(
+        self,
+        layout: SlangProgramLayout,
+        torch_dtype: torch.dtype,
+        slang_dtype: SlangType,
+        dims: int,
+        d_in: Optional["TensorRefMarshall"],
+        d_out: Optional["TensorRefMarshall"],
+    ):
+
+        dtype = innermost_type(slang_dtype)
+        can_convert = (
+            is_nested_array(slang_dtype)
+            or isinstance(slang_dtype, ScalarType)
+            or isinstance(slang_dtype, VectorType)
+            or isinstance(slang_dtype, MatrixType)
+        )
+        if not can_convert or len(slang_dtype.shape) > 2:
+            raise ValueError(f"Torch tensors do not support data type {slang_dtype.full_name}")
+
+        full_dims = dims + len(slang_dtype.shape)
+
+        super().__init__(layout, dtype, full_dims, True, d_in, d_out)
+        self.d_in: Optional[TensorRefMarshall]
+        self.d_out: Optional[TensorRefMarshall]
+
+        self.torch_dtype = torch_dtype
+        self.slang_dtype = slang_dtype
+
+    def get_shape(self, value: Optional[TensorRef] = None) -> Shape:
+        if value is not None:
+            return Shape(cast(torch.Tensor, value.tensor).shape)  # type: ignore
+        else:
+            return Shape((-1,) * self.dims)
+
+    def create_calldata(
+        self, context: CallContext, binding: "BoundVariableRuntime", data: TensorRef
+    ) -> Any:
+        if data.tensor is None:
+            raise ValueError("Missing required tensor data")
+        primal = cast(torch.Tensor, data.tensor)  # type: ignore
+
+        data.last_access = binding.access
+        shape = tuple(primal.shape)
+        strides = primal.stride()
+
+        bound_shape = shape[-len(binding.vector_type.shape) :]
+        if any([b != -1 and a != b for a, b in zip(bound_shape, binding.vector_type.shape)]):  # type: ignore
+            raise ValueError(
+                f"Tensor shape {shape} does not match expected shape {binding.vector_type.shape}"
+            )
+        assert primal.is_cuda
+
+        if context.device.info.type != DeviceType.cuda:
+
+            data_type = _torch_to_data_type[self.torch_dtype]
+            data.interop_buffer = get_storage(context, primal.numel(), primal.element_size())
+
+            interop_tensor = cast(
+                torch.Tensor,
+                data.interop_buffer.to_torch(type=data_type, shape=shape, strides=strides),
+            )
+            interop_tensor.copy_(primal)
```

Comment:
This may copy wrong data or read out of bounds if `primal` has non-standard strides. The interop buffer is created assuming a dense layout (based on `numel` and `element_size`), but the `to_torch` call passes the `strides` of the original tensor, which might not be dense.

For example:
```
a = torch.randn((64, 64), dtype=torch.float)
b = a[::2]
```
Passing `b` to slangpy causes a `CUDA_ERROR_ILLEGAL_ADDRESS`